src/main.cpp

```diff
@@ -174,19 +138,27 @@ int Main(vector<string> args)
 		
 		parser[r.c_str()].enter = [r](const char* s, size_t n, any& dt) {
 			auto& indent = *dt.get<int*>();
-			for(int i=0; i<indent; i++){cout << "|  ";}
-			string beginning(s, n);
-			if(beginning.length()>10){beginning = beginning.substr(0,10) + "...";}
-			for(auto& c:beginning){if(c == '\n'){c = '\\';}}
-			cout << r << " => \"" << beginning << "\"?" << endl;
+			cout << repeat("|  ", indent) << r << " => \"" << shorten(string(s, n), 40) << "\"?" << endl;
 			indent++;
 		};
 
-		parser[r.c_str()].leave = [r](const char* s, size_t n, bool match, any& dt) {
+		parser[r.c_str()].leave = [r, &L](const char* s, size_t n, size_t matchlen, any& value, any& dt) {
 			auto& indent = *dt.get<int*>();
 			indent--;
-			for(int i=0; i<indent; i++){cout << "|  ";}
-			cout << "`-> " << (match ? "match" : "failed") << endl;
+			cout << repeat("|  ", indent) << "`-> ";
+			if(success(matchlen)){
+
+				// display "match", the matched string and the result of the reduction
+				cout << "match: \"" << shorten(string(s, matchlen), 40) << "\" -> ";
```

Comment:
If `shorten` is always called with the pattern `shorten(string(s, len), maxlen)`, it might make sense to change the signature to `shorten(const char* s, std::size_t len, std::size_t maxlen)` to avoid gratuitous string copies.